src/main/java/software/amazon/smithy/lsp/language/FoldingRangeHandler.java

```diff
@@ -0,0 +1,198 @@
+     * trait statement for the trait block.
+     *
+     * @param trait         The first trait statement for this potential trait block
+     * @param iterator      The list iterator for the statements passed into this handler class
+     * @param foldingRanges The list holds folding ranges for current document
+     */
+    private void processFoldingRangeForTraitApplication(Syntax.Statement.TraitApplication trait, ListIterator<Syntax.Statement> iterator, List<FoldingRange> foldingRanges) {
+        int traitBlockStart = trait.start();
+        int traitBlockEnd = -1;
+        // Create folding range for the start trait statement
+        processFoldingRangeForNode(trait.value(), foldingRanges);
+        // Find next non-trait statement and create folding range for the statement traversed
+        while (iterator.hasNext()) {
+            var nextStatement = iterator.next();
+
+            if (nextStatement instanceof Syntax.Statement.TraitApplication nextTrait) {
+                traitBlockEnd = nextTrait.value() == null ? nextTrait.end() : nextTrait.value().end();
+                processFoldingRangeForNode(nextTrait.value(), foldingRanges);
+            } else {
+                iterator.previous();
+                break;
+            }
+        }
+        //Single nested trait is handled by createFoldingRangeForNode
+        if (traitBlockEnd != -1) {
+            addFoldingRange(traitBlockStart, traitBlockEnd, foldingRanges);
+        }
+    }
+
+    /**
+     * Recursively process the node's folding range for nested declarations.
+     *
+     * @param node          The node object to be processed
+     * @param foldingRanges The list holds folding ranges for current document
+     */
+    private void processFoldingRangeForNode(Syntax.Node node, List<FoldingRange> foldingRanges) {
+        if (node == null) {
+            return;
+        }
+
+        Range range = document.rangeBetween(node.start(), node.end());
+
+        if (!isFoldable(range)) {
+            return;
+        }
+
+        switch (node) {
+            case Syntax.Node.Kvps kvps -> {
+                if (kvps.kvps().isEmpty()) {
+                    return;
+                }
+
+                addFoldingRange(kvps.start(), kvps.end(), foldingRanges);
+
+                kvps.kvps().forEach(kvp -> processFoldingRangeForNode(kvp.value(), foldingRanges));
+            }
+            // Obj only contains kvps, will use kvps as its folding range
+            case Syntax.Node.Obj obj -> processFoldingRangeForNode(obj.kvps(), foldingRanges);
+
+            case Syntax.Node.Arr arr -> {
+                if (arr.elements().isEmpty()) {
+                    return;
+                }
+
+                addFoldingRange(arr.start(), arr.end(), foldingRanges);
+
+                arr.elements().forEach(element -> processFoldingRangeForNode(element, foldingRanges));
+            }
+
+            default -> addFoldingRange(node.start(), node.end(), foldingRanges);
```

Comment:
Similar to my previous comment, I wonder if this ends up creating a range. Maybe if its a string block